### PR TITLE
[go_router_builder]: ShellRoute support

### PR DESCRIPTION
**Pull request by hawkkiller**
This PR adds possibility to generate shell routes from the newest 5.0.0 API. 

flutter/flutter#111909

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by hawkkiller** [comment]
Proposal:
image
image

**Comment by hawkkiller** [comment]
@chunhtai Hi! It is my first contribution here. Looks like you've reviewed previous PR about go_router. Also, idk how to deal with this CI checks, should I do separate PRs for go_router_builder and go_router?

**Comment by hawkkiller** [comment]
image

File: packages/go_router_builder/example/lib/main.dart [hunk 1]
```diff
@@ -63,20 +64,26 @@ class App extends StatelessWidget {
   );
 }
 
+const key = GlobalObjectKey<NavigatorState>('navigator_key');
```
Comment (chunhtai) [inline packages/go_router_builder/example/lib/main.dart:67]: maybe good to create a new example instead and keep the main.dart simple

File: packages/go_router_builder/example/lib/main.g.dart [hunk 2]
```diff
@@ -49,7 +55,19 @@ extension $HomeRouteExtension on HomeRoute {
 }
 
 extension $FamilyRouteExtension on FamilyRoute {
-  static FamilyRoute _fromState(GoRouterState state) => FamilyRoute(
+  static FamilyRoute _fromState(GoRouterState state) => const FamilyRoute();
+
+  String get location => GoRouteData.$location(
```
Comment (chunhtai) [inline packages/go_router_builder/example/lib/main.g.dart:60]: shellroute should not have location, go, and push API.

File: packages/go_router_builder/lib/src/route_config.dart [hunk 3]
```diff
@@ -90,18 +96,56 @@ class RouteConfig {
     // TODO(kevmoo): validate that this MUST be a subtype of `GoRouteData`
     final InterfaceElement classElement = typeParamType.element2;
 
-    final RouteConfig value = RouteConfig._(path, classElement, parent);
+    final RouteConfig value = RouteConfig._(
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:99]: Since there is nothing to code gen the shellroute, we may be able to skip this?

**Review by chunhtai — COMMENTED** [review]
The go_router change need to go in with a separate pr first, can you open a new one and move the go_router change there?

**Comment by hawkkiller** [comment]
> The go_router change need to go in with a separate pr first, can you open a new one and move the go_router change there?

of course, but how to depend on it, so go_router_builder can see new APIs? Or when the changes approved for go_router, then we publish new library ver and after it we start implementing the go_router_builder stuff? @chunhtai

**Comment by chunhtai** [comment]
> of course, but how to depend on it, so go_router_builder can see new APIs? Or when the changes approved for go_router, then we publish new library ver and after it we start implementing the go_router_builder stuff? @chunhtai

yes, once the go_router change is merged, the go_router_builder should be able to see the change, we will need to bump the go_router version in example folder though

**Comment by hawkkiller** [comment]
> 

understood, thank you! i will do the changes in go_router and remove them from this PR.

**Comment by Skogsfrae** [comment]
@hawkkiller any update on this?

**Comment by stuartmorgan** [comment]
@hawkkiller Are you still planning on updating this PR based on the feedback above?

**Comment by hawkkiller** [comment]
@stuartmorgan @Skogsfrae 
check this https://github.com/flutter/packages/pull/2730

**Pull request closed without merging**